### DISCUSSION TITLE
Question: How can I deal with `Argument type is partially unknown` ?

### DISCUSSION THREAD

**Discussion by andyluooo** [microsoft/pyright#4855] [General]
Hi there, it might be a boring question, but I‘m asking for instructions about how to deal with the `Argument type is partially unknown` error. I'm just a beginner in Python and currently coding on VSCode with Pylance extension to help correct my mistakes. However, I am recently faced with the `Argument type is partially unknown` error and have no idea how to solve it. I've read part of the documentation but failed to find the answer since something there is too hard for me. Therefore I turn for your help and hopefully you can give me some suggestions. 
![image](https://user-images.githubusercontent.com/128155646/228235392-38e20ed1-da98-452c-bb52-a06684b75f46.png)
In this code I am trying to modify the last element of `ahha`, which is a list of integers. `left` is also an integer. This code is runnable but with a `reportUnknownArgumentType` diagnostic. I've tried to explicitly state the type of `ahha` using `ahha: list[int] = []` at the beginning of the file but didn't work. Looking forward to your reply. Thanks.

**Comment by erictraut** [comment]
Converting this to a discussion topic since it's not a bug report or enhancement request.

**Reply by erictraut** [reply]
The error you're talking about here is a strict-mode error that is not enabled by default for basic type checking. The easiest way to avoid this error is to not enable the check. 

For more specific advice, I'd need you to provide a self-contained minimal example. Diagnosing issues from a partial screen shot isn't really possible.

**Reply by andyluooo** [reply]
Maybe I can show you the code of the full function but I don't know how to upload my code correctly. I 'm just a beginner. Sorry for my awful code.  
`class Solution:
    def longestValidParentheses(self, s: str) -> int:
        foo: list[str] = []
        ahha = [0]
        mark: list[int] = []
        output: list[int] = []

        for i in range(len(s)):
            if s[i] == '(':
                right = foo.count(')')

                if right:
                    left = foo.count('(')

                    if left == right:
                        ahha[-1] += right * 2

                    elif left > right:
                        ahha.append(right * 2)
                        mark.append(left - right)

                    else:
                        ahha[-1] += left * 2    # Argument type is partially unknown
                        remain = right - left

                        while mark and remain:
                            ahha[-1] += min([remain, mark[-1]]) * 2    # Argument type is partially unknown

                            if mark[-1] > remain:
                                mark[-1] -= remain    # Argument type is partially unknown
                                remain = 0

                            else:
                                remain -= mark[-1]
                                del mark[-1]
                                ahha[-1] += ahha[-2]    # Argument type is partially unknown
                                del ahha[-2]

                        if not mark and remain:
                            output.append(ahha[0])
                            ahha = [0]

                    foo.clear()

                foo.append('(')

            elif foo:
                foo.append(')')

        left = foo.count('(')
        right = foo.count(')')

        if left == right:
            ahha[-1] += right * 2

        elif left > right:
            ahha.append(right * 2)

        else:
            ahha[-1] += left * 2
            remain = right - left

            while mark and remain:
                ahha[-1] += min([remain, mark[-1]]) * 2

                if mark[-1] > remain:
                    remain = 0

                else:
                    remain -= mark[-1]
                    del mark[-1]
                    ahha[-1] += ahha[-2]
                    del ahha[-2]

            if not mark and remain:
                output.append(ahha[0])
                ahha = [0]

        output.append(max(ahha))

        return max(output)`

**Reply by erictraut** [reply]
That code isn't self-contained. It includes a bunch of symbols (like `s` and `foo`) that are not defined in your code sample, so it doesn't help much.

If you're just beginning, then I recommend that you don't use "strict" type checking. Just stick with "basic" for now. It will detect and report the most common programming errors. Strict type checking is for developers who are writing production code and want to ensure the highest levels of robustness and maintainability.

**Reply by andyluooo** [reply]
Sorry for providing such a bad code example. Anyway, thanks for your advice. BTW, Pylance is awesome and does help me a lot on the way learning Python. I really appreciate all your efforts to maintain it. Thank you.

**Reply by Azureblade3808** [reply]
@Andy-Dihong-Luo I've fixed format of the code part for you --
```python
class Solution:
    def longestValidParentheses(self, s: str) -> int:
        foo: list[str] = []
        ahha = [0]
        mark: list[int] = []
        output: list[int] = []

        for i in range(len(s)):
            if s[i] == '(':
                right = foo.count(')')

                if right:
                    left = foo.count('(')

                    if left == right:
                        ahha[-1] += right * 2

                    elif left > right:
                        ahha.append(right * 2)
                        mark.append(left - right)

                    else:
                        ahha[-1] += left * 2    # Argument type is partially unknown
                        remain = right - left

                        while mark and remain:
                            ahha[-1] += min([remain, mark[-1]]) * 2    # Argument type is partially unknown

                            if mark[-1] > remain:
                                mark[-1] -= remain    # Argument type is partially unknown
                                remain = 0

                            else:
                                remain -= mark[-1]
                                del mark[-1]
                                ahha[-1] += ahha[-2]    # Argument type is partially unknown
                                del ahha[-2]

                        if not mark and remain:
                            output.append(ahha[0])
                            ahha = [0]

                    foo.clear()

                foo.append('(')

            elif foo:
                foo.append(')')

        left = foo.count('(')
        right = foo.count(')')

        if left == right:
            ahha[-1] += right * 2

        elif left > right:
            ahha.append(right * 2)

        else:
            ahha[-1] += left * 2
            remain = right - left

            while mark and remain:
                ahha[-1] += min([remain, mark[-1]]) * 2

                if mark[-1] > remain:
                    remain = 0

                else:
                    remain -= mark[-1]
                    del mark[-1]
                    ahha[-1] += ahha[-2]
                    del ahha[-2]

            if not mark and remain:
                output.append(ahha[0])
                ahha = [0]

        output.append(max(ahha))

        return max(output)
```

@erictraut Here's a simplified sample --
```python
any = globals()[""]

l: list[int] = any
while any:
    if any:
        l[0] += 0
    else:
        l[0] += 0  # Argument type is partially unknown.
```
Looks like a bug to me, because --
```python
any = globals()[""]

l: list[int] = any
while any:
    if any:
        l[0] += 0
        continue
    l[0] += 0  # Passes.
```
doesn't emit any error.

**Reply by andyluooo** [reply]
@Azureblade3808 Thanks for your help. This also confused me and I'm not sure if it's a bug. Hopefully someone could figure it out.

**Reply by erictraut** [reply]
Thanks for isolating the repro. Yes, this is a bug. It will be fixed in the next release of pyright.

**Reply by andyluooo** [reply]
Glad to hear that.
